code/hsec-sync/src/Security/Advisories/Sync/Snapshot.hs

```diff
@@ -0,0 +1,181 @@
+  | SnapshotDirectoryIncoherent
+
+snapshotRepositoryStatus :: Snapshot -> IO SnapshotRepositoryStatus
+snapshotRepositoryStatus s = do
+  dirExists <- D.doesDirectoryExist $ snapshotRoot s
+  if dirExists
+    then do
+      dirAdvisoriesExists <- D.doesDirectoryExist $ snapshotRoot s </> "advisories"
+      fileMetadataExists <- D.doesFileExist $ snapshotRoot s </> "snapshot.json"
+      return $
+        if dirAdvisoriesExists && fileMetadataExists
+          then SnapshotDirectoryInitialized
+          else SnapshotDirectoryIncoherent
+    else return SnapshotDirectoryMissing
+
+data SnapshotRepositoryEnsuredStatus
+  = SnapshotRepositoryCreated
+  | SnapshotRepositoryExisting
+
+ensureSnapshot ::
+  Snapshot ->
+  SnapshotRepositoryStatus ->
+  IO (Either SnapshotError SnapshotRepositoryEnsuredStatus)
+ensureSnapshot s =
+  \case
+    SnapshotDirectoryMissing ->
+      ($> SnapshotRepositoryCreated)
+        <$> overwriteSnapshot s
+    SnapshotDirectoryIncoherent ->
+      ($> SnapshotRepositoryCreated)
+        <$> overwriteSnapshot s
+    SnapshotDirectoryInitialized ->
+      return $ Right SnapshotRepositoryExisting
+
+overwriteSnapshot :: Snapshot -> IO (Either SnapshotError ())
+overwriteSnapshot s =
+  runExceptT $ do
+    let root = snapshotRoot s
+    ensuringPerformed <- liftIO $ try $ ensureEmptyRoot root
+    whenLeft ensuringPerformed $
+      throwE . SnapshotProcessError . DirectorySetupSnapshotArchive
+
+    resultE <- liftIO $ try $ get $ snapshotArchiveUrl s
+    case resultE of
+      Left e ->
+        throwE $
+          SnapshotProcessError $
+            FetchSnapshotArchive $
+              case e of
+                InvalidUrlException url reason ->
+                  "Invalid URL " <> show url <> ": " <> show reason
+                HttpExceptionRequest _ content ->
+                  case content of
+                    StatusCodeException response body ->
+                      "Request failed with " <> show (response ^. responseStatus) <> ": " <> show body
+                    _ ->
+                      "Request failed: " <> show content
+      Right result -> do
+        performed <-
+          liftIO $
+            try $
+              withSystemTempDirectory "security-advisories" $ \tempDir -> do
+                let archivePath = tempDir <> "/snapshot-export.tar.gz"
+                BL.writeFile archivePath $ result ^. responseBody
+                contents <- BL.readFile archivePath
+                Tar.unpack root $ Tar.read $ GZip.decompress contents
+        whenLeft performed $
+          throwE . SnapshotProcessError . ExtractSnapshotArchive
+
+ensureEmptyRoot :: FilePath -> IO ()
+ensureEmptyRoot root = do
+  D.createDirectoryIfMissing False root
+
+  whenM (D.doesDirectoryExist $ root </> "advisories") $
+    D.removeDirectoryRecursive $
+      root </> "advisories"
+
+  whenM (D.doesFileExist $ root </> "snapshot.json") $
+    D.removeFile $
+      root </> "snapshot.json"
```

Comment:
I think this is a dangerous thing, just removing the files without asking.

Reply:
especially the `removeDirectoryRecursive`

Reply:
It's the purpose of overwriting a snapshot, asking to details at this point would add complexity while we only expect to have a function which takes care of everything.

Reply:
Okay, perhaps let me reformulate, is it possible that we ever delete something here that the user might want to keep? If that’s the case, we should make it opt in, I think.

Reply:
But perhaps I’m also overthinking. ;)